File: android/app/src/main/kotlin/com/goalmaker/app/ui/connector/ConnectorViewModel.kt
package com.goalmaker.app.ui.connector

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.goalmaker.app.application.connector.ConnectorLinks
import com.goalmaker.app.application.sync.RemoteRejectedException
import com.goalmaker.app.application.sync.RemoteUnavailableException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Creates, rotates and revokes the Claude connector's link (docs/connector.md). A new link's URL is
 * kept only in this screen's state, shown once, and never written anywhere.
 */
class ConnectorViewModel(
    private val links: ConnectorLinks,
    private val backendUrl: String,
    private val io: CoroutineDispatcher,
) : ViewModel() {
    private val state = MutableStateFlow(ConnectorUiState())
    val uiState: StateFlow<ConnectorUiState> = state.asStateFlow()

    // The link the new URL belongs to, so the URL goes once that link stops being the active one.
    private var newLinkId: String? = null

    init {
        refresh()
    }

    fun refresh() = call { withContext(io) { load() } }

    /** Makes a link; when one is active, this is rotating: the old one stops working. */
    fun create() = call {
        val secret = withContext(io) { links.create() }
        newLinkId = null
        state.update { it.copy(newUrl = ConnectorLinks.url(backendUrl, secret)) }
        withContext(io) { load() }
    }

    fun revoke() = call {
        withContext(io) { links.revoke() }
        state.update { it.copy(newUrl = null) }
        withContext(io) { load() }
    }

    /** The owner copied the new link; it isn't shown again. */
    fun hideNewLink() = state.update { it.copy(newUrl = null) }

    private suspend fun load() {
        val active = links.list().firstOrNull { it.active }
        if (state.value.newUrl != null) {
            // Made just now: remember its link. Revoked or replaced elsewhere since: the URL is dead.
            if (newLinkId == null) newLinkId = active?.id
            if (active == null || active.id != newLinkId) state.update { it.copy(newUrl = null) }
        }
        state.update { it.copy(loaded = true, active = active, unavailable = false) }
    }

    private fun call(work: suspend () -> Unit) {
        if (state.value.busy) return
        viewModelScope.launch {
            state.update { it.copy(busy = true) }
            try {
                work()
            } catch (_: RemoteUnavailableException) {
                state.update { it.copy(loaded = true, unavailable = true) }
            } catch (_: RemoteRejectedException) {
                state.update { it.copy(loaded = true, unavailable = true) }
            } finally {
                state.update { it.copy(busy = false) }
            }
        }
    }
}

File: windows/src/GoalMaker.App/ViewModels/ConnectorViewModel.cs
using System.Globalization;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using GoalMaker.App.Localization;
using GoalMaker.Core.Connector;
using GoalMaker.Core.Sync;

namespace GoalMaker.App.ViewModels;

/// <summary>
/// The Claude connector card in Settings (docs/connector.md): the link's state, Create, Make a new link
/// and Revoke, each of the last two confirmed in place because they stop the current link. A new link's
/// URL lives only here, shown once, and is never written anywhere.
/// </summary>
public sealed partial class ConnectorViewModel : ObservableObject
{
    private readonly IConnectorLinks links;
    private readonly string backendUrl;
    private readonly IStrings strings;
    private readonly Action<string> copy;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowCreate), nameof(ShowActive))]
    private bool loaded;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowCreate), nameof(ShowActive))]
    private bool unavailable;

    [ObservableProperty]
    [NotifyCanExecuteChangedFor(nameof(CreateCommand), nameof(RotateCommand), nameof(RevokeCommand), nameof(ConfirmCommand), nameof(RefreshCommand))]
    private bool busy;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowCreate), nameof(ShowActive))]
    private bool hasActive;

    [ObservableProperty]
    private string status = string.Empty;

    [ObservableProperty]
    private string lastUsed = string.Empty;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(HasNewUrl))]
    private string newUrl = string.Empty;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(IsConfirming), nameof(ConfirmText), nameof(ConfirmLabel))]
    private ConnectorAction? pending;

    [ObservableProperty]
    private bool copied;

    public ConnectorViewModel(IConnectorLinks links, string backendUrl, IStrings strings, Action<string> copy)
    {
        this.links = links;
        this.backendUrl = backendUrl;
        this.strings = strings;
        this.copy = copy;
    }

    public bool HasNewUrl => NewUrl.Length > 0;

    /// <summary>No link yet: offer to make one.</summary>
    public bool ShowCreate => Loaded && !Unavailable && !HasActive;

    /// <summary>A link is active: say when it was made and used, and offer a new one or revoking it.</summary>
    public bool ShowActive => Loaded && !Unavailable && HasActive;

    public bool IsConfirming => Pending is not null;

    public string ConfirmText => Pending switch
    {
        ConnectorAction.Rotate => strings.Get("Connector.RotateWarning"),
        ConnectorAction.Revoke => strings.Get("Connector.RevokeWarning"),
        _ => string.Empty,
    };

    public string ConfirmLabel => Pending == ConnectorAction.Revoke ? strings.Get("Connector.Revoke") : strings.Get("Connector.Rotate");

    /// <summary>Reads the link's state; the page calls it when it opens.</summary>
    [RelayCommand(CanExecute = nameof(CanCall))]
    public Task RefreshAsync() => CallAsync(LoadAsync);

    private bool CanCall() => !Busy;

    [RelayCommand(CanExecute = nameof(CanCall))]
    private Task CreateAsync() => CallAsync(async () =>
    {
        var secret = await links.CreateAsync();
        NewUrl = IConnectorLinks.Url(backendUrl, secret);
        Copied = false;
        await LoadAsync();
    });

    [RelayCommand(CanExecute = nameof(CanCall))]
    private void Rotate() => Pending = ConnectorAction.Rotate;

    [RelayCommand(CanExecute = nameof(CanCall))]
    private void Revoke() => Pending = ConnectorAction.Revoke;

    [RelayCommand(CanExecute = nameof(CanCall))]
    private Task ConfirmAsync()
    {
        var action = Pending;
        Pending = null;
        return action == ConnectorAction.Revoke
            ? CallAsync(async () =>
            {
                await links.RevokeAsync();
                NewUrl = string.Empty;
                await LoadAsync();
            })
            : CreateAsync();
    }

    [RelayCommand]
    private void Cancel() => Pending = null;

    [RelayCommand]
    private void Copy()
    {
        if (HasNewUrl)
        {
            copy(NewUrl);
            Copied = true;
        }
    }

    /// <summary>The owner is done with the new link; it isn't shown again.</summary>
    [RelayCommand]
    private void HideNewUrl()
    {
        NewUrl = string.Empty;
        Copied = false;
    }

    private async Task LoadAsync()
    {
        var active = (await links.ListAsync()).FirstOrDefault(link => link.Active);
        HasActive = active is not null;
        Status = active is null ? strings.Get("Connector.None") : strings.Get("Connector.Active", Format(active.CreatedAt));
        LastUsed = active is null
            ? string.Empty
            : active.LastUsedAt is { } used ? strings.Get("Connector.LastUsed", Format(used)) : strings.Get("Connector.NeverUsed");
        Unavailable = false;
        Loaded = true;
    }

    private async Task CallAsync(Func<Task> work)
    {
        if (Busy)
        {
            return;
        }

        Busy = true;
        try
        {
            await work();
        }
        catch (Exception error) when (error is RemoteUnavailableException or RemoteRejectedException)
        {
            Unavailable = true;
            Loaded = true;
        }
        finally
        {
            Busy = false;
        }
    }

    private static string Format(DateTimeOffset at) => at.ToLocalTime().ToString("ddd d MMM, HH:mm", CultureInfo.CurrentCulture);
}
